zkevm-circuits/src/evm_circuit/util/common_gadget.rs

```diff
@@ -815,3 +815,99 @@ pub(crate) fn cal_sstore_gas_cost_for_assignment(
         warm_case_gas.0 + GasCost::COLD_SLOAD.0
     }
 }
+
+#[derive(Clone, Debug)]
+pub(crate) struct CommonErrorGadget<F> {
+    rw_counter_end_of_reversion: Cell<F>,
+    restore_context: RestoreContextGadget<F>,
+}
+
+impl<F: Field> CommonErrorGadget<F> {
+    pub(crate) fn construct(
+        cb: &mut ConstraintBuilder<F>,
+        opcode: Expression<F>,
+        rw_counter_delta: Expression<F>,
+    ) -> Self {
+        cb.opcode_lookup(opcode.expr(), 1.expr());
+
+        let rw_counter_end_of_reversion = cb.query_cell();
+
+        // current call must be failed.
+        cb.call_context_lookup(false.expr(), None, CallContextFieldTag::IsSuccess, 0.expr());
+
+        cb.call_context_lookup(
+            false.expr(),
+            None,
+            CallContextFieldTag::RwCounterEndOfReversion,
+            rw_counter_end_of_reversion.expr(),
+        );
+
+        // Go to EndTx only when is_root
+        let is_to_end_tx = cb.next.execution_state_selector([ExecutionState::EndTx]);
+        cb.require_equal(
+            "Go to EndTx only when is_root",
+            cb.curr.state.is_root.expr(),
+            is_to_end_tx,
+        );
+
+        // When it's a root call
+        cb.condition(cb.curr.state.is_root.expr(), |cb| {
+            // Do step state transition
+            cb.require_step_state_transition(StepStateTransition {
+                call_id: Same,
+                rw_counter: Delta(rw_counter_delta + cb.curr.state.reversible_write_counter.expr()),
+                ..StepStateTransition::any()
+            });
+        });
+
+        // When it's an internal call, need to restore caller's state as finishing this
+        // call. Restore caller state to next StepState
+        let restore_context = cb.condition(1.expr() - cb.curr.state.is_root.expr(), |cb| {
+            RestoreContextGadget::construct(
```

Comment:
I think it doesn't make much sense to add newn associated fn for `RestoreContextGadget` like: `RestoreContextGadget::from_revert_ctx` which directly places all the 0's and has less "magic numbers" arround.
as after refactoring, it is only one place to use it now. WDYT? @CPerezz

Reply:
Makes sense! I like it that way!